lambda/simplif.ml

```diff
@@ -491,13 +499,21 @@ let simplify_lets lam =
 (* This (small)  optimisation is always legal, it may uncover some
    tail call later on. *)
 
-  let mklet str kind v e1 e2  = match e2 with
-  | Lvar w when optimize && Ident.same v w -> e1
-  | _ -> Llet (str, kind,v,e1,e2) in
+  let mklet str kind v e1 e2 =
+    match e2 with
+    | Lvar w when optimize && Ident.same v w -> e1
+    | _ -> Llet (str, kind,v,e1,e2)
+  in
 
+  let mkmutlet kind v e1 e2 =
+    match e2 with
+    | Lmutvar w when optimize && Ident.same v w -> e1
+    | _ -> Lmutlet (kind,v,e1,e2)
+  in
```

Comment:
I believe that this optimization is not correct. Consider:

```
let-mut x = 0 in
ignore (let-mut x = x in x <- 1);
x (* I expect 0, the optimization would give 1 *)
```

Reply:
`let-mut x = x in x` is not a valid left-hand side for `Lassign`. So this part of your example actually looks like `let-mut x = x in (x <- 1)`, which does not trigger the optimisation.
That said, to get the optimisation to trigger you would need to write source code in the form `let x = ref foo in !x`. I don't think it's common enough to justify keeping it around, so I'm fine with removing it, even if it would technically be a regression.

Reply:
Ah, apologies, I got confused again between optimizing `let x = x in e` and `let x = e in x`. I had thought about this during my first pass reading the code, and indeed there is no problem with `let-mut x = e in x`, and my previous message should be ignored.